src/util.ts

```diff
@@ -130,3 +130,48 @@ export function joinAll(a: ArrayBuffer[]): ArrayBuffer {
 
     return buffer;
 }
+
+// Consumes data:
+//   PrivateToken challenge="abc...", token-key="123...",
+//   PrivateToken challenge="def...", token-key="234...",
+//   PrivateToken challenge=ghi..., token-key=345...
+// Parse WWW-Authenticate according to RFC9110 Section 11.6.1 https://www.rfc-editor.org/rfc/rfc9110#section-11.6.1
+export function parseWWWAuthenticate(header: string): string[] {
+    const ALPHA = 'A-Za-z';
+    const DIGIT = '0-9';
+    const tokenChar = `!#$%&'*+\\-\\.^_\`|~${DIGIT}${ALPHA}`;
+    const tchar = '[' + tokenChar + ']';
+    const OWS = '[ \\t]*';
+    const RWS = '[ \\t]+';
+    const BWS = OWS;
+    const qdtext = '[\\t\\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]';
+    const quotedPair = '\\\\([ \\t' + tokenChar + '\\x21-\\x7E\\x80-\\xFF])';
+    const quotedString = `"(?:${qdtext}|${quotedPair})*"`;
+    const token = tchar + '+';
+    const authParam = `${token}${BWS}=${BWS}(?:${token}|${quotedString})`;
+    const authScheme = token;
+    const token68 = `[\\/+\\-\\._~${DIGIT}${ALPHA}]+=*`;
+    const challenge = `${authScheme}(?:${RWS}(?:${token68}|${authParam}(?:${OWS},${OWS}${token68}|${authParam})*))?`;
+    const challenges = `(?<challenge>${challenge})(?<to_remove>${OWS},${OWS}(?<challenge_optional>${challenge}))*`;
+
+    const challengesRegex = new RegExp(`^${challenges}$`);
+
+    const challengesRaw: string[] = [];
+    let toMatch = header;
+    do {
+        const groups = toMatch.match(challengesRegex)?.groups;
+        if (!groups) {
+            return [];
+        }
+        if (groups.challenge_optional) {
+            challengesRaw.push(groups.challenge_optional);
```

Comment:
This seems rather expensive, matching the whole thing over and over again, throwing away the result.

Here is a sketch of a simplified alternative that does a single pass using a [sticky](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/sticky) RegExp flag to ensure that no substring is ignored.
```javascript
let testCases = [
    ["A, B, C", ["A", "B", "C"]],
    ["A,  B, C", ["A", "B", "C"]],
    ["A,B", ["A", "B"]],
    ["A", ["A"]],
    ["A,", []],
    ["A,Bx", []],
    [",A", []],
];

testCases.forEach(([str, expected]) => {
    let regex = /(?<skip> *, *)?(?<data>[A-Z]+)/y;
    let first = true;
    var m;
    let everythingConsumed = false;
    let matches = [];
    while ((m = regex.exec(str))) {
        if (first) {
            if (m.groups.skip) {
                break;
            }
            first = false;
        }
        matches.push(m.groups.data);
        everythingConsumed = str.length === regex.lastIndex;
    }
    if (!everythingConsumed) {
        matches = [];
    }
    //console.log(matches);
    if (matches.toString() !== expected.toString()) {
        console.log("expected", expected, "got", matches);
    }
});
```

Reply:
updated
I've slightly modified the naming and the while to scope `m` variable to the for loop.